### PR TITLE
Improve layout and formatting of `concept.rst` for GUI system docs

### PR DESCRIPTION
**Pull request by eschan145**
Added a lot of documentation and a good tutorial on the GUI concepts. Also added more reStructuredText formatting.

Pyglet spells its own name with a lowercase first letter, so I have included it here. I also added documentation on styling here, maybe later that might be moved to `style.rst`. It's easier to visualize when you have style and widgets next to each other.

Also moved some text (such as parameters, class names) into code blocks to make it look neater. Please let me know if you have any suggestions.

### CODE REVIEW

File: doc/programming_guide/gui/concept.rst [hunk 1]
```diff
@@ -3,79 +3,99 @@
...
+identify and search for in autocompletion.
 
 UIWidget
 ========
 
-:class:`UIWidget` are the core of Arcades GUI. A widget represents the behaviour and graphical
-representation of any element (like Buttons or Text)
+A :py:class:`~arcade.class.UIWidget` is the core of arcade's GUI. A widget
+represents the behaviour and graphical representation of any element (like
+buttons or text).
```
Comment (pushfoo) [inline doc/programming_guide/gui/concept.rst]: 1. "A" is inappropriate here since the concept being referred to is the class, not the instances of them.
2. `arcade.class.UIWidget` does not appear to be a valid module
3. This should probably rephrased to remove ambiguity in a number of areas.

```suggestion
The :py:class:`~arcade.gui.UIWidget` class is the core of arcade's GUI system. Widgets
specify the behavior and graphical representation of any UI element, such as buttons
or labels.
```

File: doc/programming_guide/gui/concept.rst [hunk 2]
```diff
@@ -3,79 +3,99 @@
...
+``size_hint_min``
+    A tuple of two integers. Defines minimal size of the widget. Changing the
+    size of a widget to a lower values will use this size instead.
 
-**size_hint_max**
-    tuple of two ints, defines maximum size of the widget.
-    If set, changing the size of a widget to a higher values will use this size instead.
+``size_hint_max``
+    A tuple of two integers. Defines maximum size of the widget. Changing the
+    size of a widget to a higher values will use this size instead.
```
Comment (pushfoo) [inline doc/programming_guide/gui/concept.rst]:
```suggestion
    A tuple of two integers defining the maximum width and height of the widget.
    Attempting to set a larger width or height greater will fail by defaulting to the
    to the maximum values specified here.
 ```

File: doc/programming_guide/gui/concept.rst [hunk 3]
```diff
@@ -3,79 +3,99 @@
...
+    A tuple of two floats. Defines how much of the parent's space it would like
+    to occupy (range: ``0.0-1.0``). For maximal vertical and horizontal
+    expansion, define a ``size_hint`` of 1 for the axis.
 
-**size_hint_min**
-    tuple of two ints, defines minimal size of the widget.
-    If set, changing the size of a widget to a lower values will use this size instead.
+``size_hint_min``
+    A tuple of two integers. Defines minimal size of the widget. Changing the
+    size of a widget to a lower values will use this size instead.
```
Comment (pushfoo) [inline doc/programming_guide/gui/concept.rst]:
```suggestion
    A tuple of two integers defining the minimum width and height of the widget.
    Attempting to set a smaller width or height on the widget will fail by defaulting
    to the minimum values specified here.
```

File: doc/programming_guide/gui/concept.rst [hunk 4]
```diff
@@ -3,79 +3,99 @@
...
+    move or resize its children; use a :py:class:`~arcade.gui.UILayout`
+    instead.
 
-**size_hint**
-    tuple of two floats, defines how much of the parents space it would like to occupy (range: 0.0-1.0).
-    For maximal vertical and horizontal expansion, define `size_hint` of 1 for the axis.
+``size_hint``
+    A tuple of two floats. Defines how much of the parent's space it would like
+    to occupy (range: ``0.0-1.0``). For maximal vertical and horizontal
+    expansion, define a ``size_hint`` of 1 for the axis.
```
Comment (pushfoo) [inline doc/programming_guide/gui/concept.rst]: We can be even clearer with examples.

```suggestion
    A tuple of two normalized floats (``0.0``-``1.0``) describing the portion of the parent's width
    and height this widget prefers to occupy.
    
    Examples::
    
        # Prefer to take up all space within the parent
        widget.size_hint = (1.0, 1.0)
    
        # Prefer to take up the full width & half the height of the parent
        widget.size_hint = (1.0, 0.5)

        # Prefer using 1/10th of the available width & height
        widget.size_hint = (0.1, 0.1)
 ```

File: doc/programming_guide/gui/concept.rst [hunk 5]
```diff
@@ -3,79 +3,99 @@
...
-    If set, changing the size of a widget to a higher values will use this size instead.
+``size_hint_max``
+    A tuple of two integers. Defines maximum size of the widget. Changing the
+    size of a widget to a higher values will use this size instead.
 
-    *size_hint*, *size_hint_min*, and *size_hint_max* are values that are additional information of a widget, but do not
-    effect the widget on its own. :class:`UILayout` may use these information to place or resize a widget.
+``size_hint``, ``size_hint_min``, and ``size_hint_max`` are values that are
+additional information of a widget, but do not effect the widget on its own.
+:class:`UILayout` may use this data to place or resize a widget.
```
Comment (pushfoo) [inline doc/programming_guide/gui/concept.rst]: This paragraph is very unclear in its current form in addition to typos (effect instead of affect). We may as well improve it.
```suggestion
.. warning:: Size hints do nothing on their own!

             They are hints to :class:`UILayout` instances, which may
             choose to use or ignore them.
```

File: doc/programming_guide/gui/concept.rst [hunk 6]
```diff
@@ -127,99 +147,317 @@ Size hint support
...
-- :class:`UILayout`:
-    - :class:`UIBoxLayout` - Places widgets next to each other (vertical or horizontal)
-    - :class:`UIAnchorLayout` - Places widgets within itself following anchor information
-    - :class:`UIGridLayout` - Places widgets within a grid
-- Constructs
-    - :class:`UIMessageBox` - Popup box with a message text and a few buttons.
-- Mixins
-    - :class:`UIDraggableMixin` - Makes a widget draggable.
-    - :class:`UIMouseFilterMixin` - Catches mouse events that occure within the widget boundaries.
-    - :class:`UIWindowLikeMixin` - Combination of :class:`UIDraggableMixin` and :class:`UIMouseFilterMixin`.
```
Comment (pushfoo) [inline doc/programming_guide/gui/concept.rst:158]: I'm not sure removing this entirely is a good idea. The sidebar menu doesn't link the sections below, so this may be useful if we keep it and build on the API link style around lines 374-396, but also add links to relevant headings within this page.

@eruvanos your thoughts?  Maybe something like this but with better line length?
```
- :py:class:`~arcade.gui.UIWidget`:
    - :py:class:`~arcade.gui.UIFlatButton` - `(concept) <pg_gui_concepts_uiflatbutton>`_ 2D flat button for simple interactions (hover, press, release, click)

...

.. pg_gui_concepts_uiflatbutton:

Flat button
^^^^^^^^^^^
```
The programming guide link anchor convention is unfinalized and only currently used in #1620 as far as I know. Regardless of what standard we settle on, my goals are:
1. Keep object docstrings simple, favoring examples
2. Link relevant programming guide pages from them

File: doc/programming_guide/gui/concept.rst [hunk 7]
```diff
@@ -127,99 +147,317 @@ Size hint support
...
-
-UIEvents are fully typed dataclasses, which provide information about a event effecting the UI.
-
-All pyglet window events are converted by the UIManager into UIEvents and passed via dispatch_event
-to the ``on_event`` callbacks.
+User-interface events
+=====================
+
+Arcade's GUI events are fully typed dataclasses, which provide information
+about a event effecting the UI.
```
Comment (pushfoo) [inline doc/programming_guide/gui/concept.rst]: 1. "an", due to event starting with a vowel sound
2. "affecting" rather than "effecting". I know this is annoying, but in english these have completely different meanings despite sounding the same.
 
```suggestion
about an event affecting the UI.
```

**Review by pushfoo — CHANGES_REQUESTED** [review]
tl;dr: good job overall. I have concerns about removing a section, made grammar / clarity suggestions in parts crucial to beginners

I made some specific phrasing tweak suggestions to fix grammar & clarity of beginner-relevant sections, but I'll leave out suggestions for the Rendering and UILayout section because I don't think I understand them well enough.

We may want to keep the hierarchical listing of UIWidgets because the titles do not show up in the sidebar. Adding link targets to each heading we could link from the overview would set us up for improving docstrings down the line in the same way einarf has approved of in #1620.

**Comment by pvcraven** [comment]
How are we feeling about this PR? I know there are some open conversations, but it looks like this is still an improvement as-is.

File: doc/programming_guide/gui/concept.rst (hunk 6 above)
Comment (eruvanos) [reply doc/programming_guide/gui/concept.rst:158]: I see your points, right now I would like to continue and merge the general improvements. We can at it latter if we feel it is really missing.

**Review by eruvanos — APPROVED** [review]
Thanks for your work, beside some open points, which were mentioned by @pushfoo it is a grade improvement of the current docs. As mentioned, we can address those points later.

File: doc/programming_guide/gui/concept.rst (hunk 1 above)
Comment (eruvanos) [reply doc/programming_guide/gui/concept.rst]: Will apply after merge
